ios/browser/web/web_state/brave_web_state_native.mm

```diff
@@ -0,0 +1,318 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/ios/browser/web/web_state/brave_web_state_native.h"
+
+#include "ios/chrome/browser/browser_state/chrome_browser_state.h"
+#include "ios/chrome/browser/main/browser.h"
+#include "ios/chrome/browser/sessions/ios_chrome_session_tab_helper.h"
+#include "ios/chrome/browser/tabs/synced_window_delegate_browser_agent.h"
+#include "ios/chrome/browser/web_state_list/web_state_list.h"
+#include "ios/chrome/browser/web_state_list/web_state_opener.h"
+
+#include "ios/web/public/session/crw_navigation_item_storage.h"
+#include "ios/web/public/session/crw_session_storage.h"
+#include "ios/web/public/thread/web_thread.h"
+#include "ios/web/public/web_state_observer.h"
+#include "ios/web/web_state/ui/crw_web_view_navigation_proxy.h"
+#include "ios/web/web_state/web_state_impl.h"
+
+#include "net/base/mac/url_conversions.h"
+#include "url/gurl.h"
+
+#if !defined(__has_feature) || !__has_feature(objc_arc)
+#error "This file requires ARC support."
+#endif
+
+#pragma mark - BackForwardList
+
+// Back Forward List for use in Navigation Manager
+@interface BackForwardList : NSObject
+@property(nonatomic, readonly, copy) WKBackForwardListItem* currentItem;
+@property(nonatomic, readonly, copy) NSArray<WKBackForwardListItem*>* backList;
+@property(nonatomic, readonly, copy)
+    NSArray<WKBackForwardListItem*>* forwardList;
+@end
+
+@implementation BackForwardList
+@synthesize currentItem;
+@synthesize backList;
+@synthesize forwardList;
+
+- (instancetype)init {
+  self = [super init];
+  return self;
+}
+
+- (WKBackForwardListItem*)itemAtIndex:(NSInteger)index {
+  if (index == 0) {
+    return currentItem;
+  }
+
+  if (index > 0 && forwardList.count) {
+    return forwardList[index - 1];
+  }
+
+  if (backList.count) {
+    return backList[backList.count + index];
+  }
+
+  return nullptr;
+}
+
+- (WKBackForwardListItem*)backItem {
+  return backList.lastObject;
+}
+
+- (WKBackForwardListItem*)forwardItem {
+  return forwardList.firstObject;
+}
+@end
+
+#pragma mark - NavigationProxy
+
+@interface NavigationProxy : NSObject <CRWWebViewNavigationProxy>
+@end
+
+@implementation NavigationProxy
+- (instancetype)init {
+  if ((self = [super init])) {
+  }
+  return self;
+}
+
+- (NSURL*)URL {
+  return net::NSURLWithGURL(GURL(url::kAboutBlankURL));
+}
+
+- (WKBackForwardList*)backForwardList {
+  return (WKBackForwardList*)[[BackForwardList alloc] init];
+}
+@end
+
+#pragma mark - NavigationDelegate
+
+class NavigationDelegate : public web::NavigationManagerDelegate {
```

Comment:
Do we need to namespace this?

Reply:
Not really since it's in the `.mm` file and never exposed, but I did it anyway.